src/write.jl

```diff
@@ -2,52 +2,116 @@
-        if length(str) <= 8
-            result = str
-        end
-    else
-        fpart, ipart = modf(x)
-        ipart_str = string('-'^signbit(x), Int(abs(ipart))) # handles `-0.0` case
-        fpart_str = @sprintf "%.7f" abs(fpart)
-        fpart_str = fpart_str[3:end] # remove leading `0.`
-        if length(ipart_str) < 7
-            result = ipart_str * '.' * fpart_str[1:(7 - length(ipart_str))]
-        elseif length(ipart_str) <= 8
-            result = ipart_str
-        end
-    end
-    if !ismissing(result)
-        if all(c -> c in ('0', '.', '-'), result)
-            x == 0 && return result
+# `allow_scientific` is only meaningful for `value::Number`. We allow passing it though,
+# so `edf_write` can be more generic.
+_edf_repr(value::Union{String,Char}; allow_scientific=nothing) = value
+_edf_repr(date::Date; allow_scientific=nothing) = uppercase(Dates.format(date, dateformat"dd-u-yyyy"))
+_edf_repr(date::DateTime; allow_scientific=nothing) = Dates.format(date, dateformat"dd\.mm\.yyHH\.MM\.SS")
+
+"""
+    sprintf_G_under_8(x) -> String
+
+Return a string of length at most 8, written in either scientific notation (using capital-'E'),
+or in decimal format, using as much precision as possible.
+"""
+function sprintf_G_under_8(x)
+    shorten = str -> begin
+        if contains(str, 'E')
+            # Remove extraneous 0's in exponential notation
+            str = replace(str, "E-0" => "E-")
+            str = replace(str, "E+0" => "E+")
+            # Remove any leading/trailing whitespace
+            return strip(str)
         else
-            return result
+            # Decimal format. Call out to `trim_F`
+            return trim_F(str)
         end
     end
+    # Strategy:
+    # `@printf("%0.NG", x)` means:
+    # - `G`: Use the shortest representation: %E or %F. That is, scientific notation (with capital E) or decimals, whichever is shorted
+    # - `.N`: (for literal `N`, like `5`): the maximum number of significant digits to be printed
+    # However, `@printf("%0.NG", x)` may have more than `N` characters (e.g. presence of E, and the values for the exponent, the decimal place, etc)
+    # So we start from 8 (most precision), and stop as soon as we get under 8 characters
+    sig_8 = shorten(@sprintf("%.8G", x))
+    length(sig_8) <= 8 && return sig_8
+    sig_7 = shorten(@sprintf("%.7G", x))
+    length(sig_7) <= 8 && return sig_7
+    sig_6 = shorten(@sprintf("%.6G", x))
+    length(sig_6) <= 8 && return sig_6
+    sig_5 = shorten(@sprintf("%.5G", x))
+    length(sig_5) <= 8 && return sig_5
+    sig_4 = shorten(@sprintf("%.4G", x))
+    length(sig_4) <= 8 && return sig_4
+    sig_3 = shorten(@sprintf("%.3G", x))
+    length(sig_3) <= 8 && return sig_3
+    sig_2 = shorten(@sprintf("%.2G", x))
+    length(sig_2) <= 8 && return sig_2
+    sig_1 = shorten(@sprintf("%.1G", x))
+    length(sig_1) <= 8 && return sig_1
+    error("failed to fit number into EDF's 8 ASCII character limit: $x")
+end
```

Comment:
These are technically Printf internals, although I agree it is much much nicer code. This will also be less performant bc we dynamically construct the format strings (which are actually expensive), but we could hoist those to a global const

Reply:
They're not in the manual (which for Printf hasn't been touched in many years) but they have docstrings and are included in the [NEWS.md entry for 1.6](https://github.com/JuliaLang/julia/blob/release-1.6/NEWS.md#printf) and are used in [a number of packages](https://juliahub.com/ui/Search?type=symbols&q=format&u=use&t=function). Good point about them being expensive though—with a global const you could then do
```julia
const FORMATS = (; G=[Printf.Format("%.$(i)G") for i in 8:-1:1],
                 F=[Printf.Format("%.$(i)F") for i in 8:-1:1])

function _edf_repr(x::Real, allow_scientific::Bool)
    fmts = allow_scientific ? FORMATS.G : FORMATS.F
    for fmt in fmts
        str = Printf.format(fmt, x)
        str = strip(replace(str, r"(E[+-])0" => s"\1", r"\.0+$" => ""))
        length(str) <= 8 && return String(str)
    end
    throw(ArgumentError("cannot represent $x in 8 ASCII characters"))
end
```

Reply:
Ok, I will PR the manual and switch to this